File: Lista4/Golang/lib/monitor.go
package lib

import "context"

type Monitor struct {
	S        ChannelSemaphore
	Fork     []int
	OKtoEatS []ChannelSemaphore
	OKtoEatC []int

	n int
}

func NewMonitor(ctx context.Context, n int) Monitor {
	m := Monitor{}
	m.S = NewChannelSemaphore(ctx, 1, n)
	m.Fork = make([]int, n)
	m.OKtoEatS = make([]ChannelSemaphore, n)
	m.OKtoEatC = make([]int, n)
	for i := 0; i < n; i++ {
		m.Fork[i] = 2
		m.OKtoEatS[i] = NewChannelSemaphore(ctx, 0, 1)
		m.OKtoEatC[i] = 0
	}
	m.n = n
	return m
}

func (m *Monitor) TakeFork(i int) {
	m.S.Wait()
	if m.Fork[i] != 2 {
		m.OKtoEatC[i] += 1
		m.S.Signal()
		m.OKtoEatS[i].Wait()
		m.S.Wait()
		m.OKtoEatC[i] -= 1
	}
	m.Fork[(i+1)%m.n] -= 1
	m.Fork[(m.n+i-1)%m.n] -= 1
	m.S.Signal()
}

func (m *Monitor) ReleaseFork(i int) {
	m.S.Wait()
	m.Fork[(i+1)%m.n] += 1
	m.Fork[(m.n+i-1)%m.n] += 1
	if m.Fork[(i+1)%m.n] == 2 {
		if m.OKtoEatC[(i+1)%m.n] != 0 {
			m.OKtoEatS[(i+1)%m.n].Signal()
		}
	}
	if m.Fork[(m.n+i-1)%m.n] == 2 {
		// ok to eat[(m.n + i-1) % m.n] signal
		if m.OKtoEatC[(m.n+i-1)%m.n] != 0 {
			m.OKtoEatS[(m.n+i-1)%m.n].Signal()
		}
	}
	m.S.Signal()
}

File: Lista4/C++/monitor.cpp
#include "monitor.hpp"

Monitor::Monitor(int numberOfPhilosophers, int numberOfMeals) {
    this->numberOfPhilosophers = numberOfPhilosophers;
    okToEatS = std::vector< std::binary_semaphore * >{static_cast<unsigned long>(numberOfPhilosophers)};
    for(int i = 0; i < numberOfPhilosophers; i++) {
        Fork.push_back(2);
        okToEatS[i] = new std::binary_semaphore{0};
        okToEatC.push_back(0);
        roster.push_back(false);
        mealsLeft.push_back(numberOfMeals);
    }
}

void Monitor::takeFork(int i) {
    int forkToTheLeft = (i + 1) % numberOfPhilosophers;
    int forkToTheRight = (numberOfPhilosophers + i - 1) % numberOfPhilosophers;
    S.acquire();
    if (Fork[i] != 2) {
        okToEatC[i] += 1;
        S.release();
        okToEatS[i]->acquire();
        S.acquire();
        okToEatC[i] -= 1;
    }
    Fork[forkToTheLeft] -= 1;
    Fork[forkToTheRight] -= 1;
    addToRoster(i);
    S.release();
}

void Monitor::releaseFork(int i) {
    int forkToTheLeft = (i + 1) % numberOfPhilosophers;
    int forkToTheRight = (numberOfPhilosophers + i - 1) % numberOfPhilosophers;
    S.acquire();
    Fork[forkToTheLeft] += 1;
    Fork[forkToTheRight] += 1;
    if (Fork[forkToTheLeft] == 2) {
        if (okToEatC[forkToTheLeft] != 0) {
            okToEatS[forkToTheLeft]->release();
        }
    }
    if (Fork[forkToTheRight] == 2) {
        if (okToEatC[forkToTheRight] != 0) {
            okToEatS[forkToTheRight]->release();
        }
    }
    removeFromRoster(i);
    S.release();
}

void Monitor::addToRoster (int i) {
    roster[i] = true;
    std::cout << "=============" << std::endl;
    std::cout << "Philosopher " << i << " started eating" << std::endl;
    if(roster[0]) {
        std::cout << "(0";
    } else {
        std::cout << "(_";
    }
    for(int j = 1; j <  numberOfPhilosophers; j++) {
        if(roster[j]) {
            std::cout << ", " << j;
        } else {
            std::cout << ", _";
        }
    }
    std::cout << ")" << std::endl;
}

void Monitor::removeFromRoster (int i) {
    roster[i] = false;
    mealsLeft[i] -=1;
    std::cout << "=============" << std::endl;
    std::cout << "Philosopher " << i << " stopped eating" << std::endl;
    if (mealsLeft[i] > 0) {
        std::cout << "Meals left: " << i << std::endl;
    } else {
        std::cout << "Philosopher " << i << " finished all of their meals" << std::endl;
    }
    if(roster[0]) {
        std::cout << "(0";
    } else {
        std::cout << "(_";
    }
    for(int j = 1; j <  numberOfPhilosophers; j++) {
        if(roster[j]) {
            std::cout << ", " << j;
        } else {
            std::cout << ", _";
        }
    }
    std::cout << ")" << std::endl;
}
